onnx/defs/function.cc

```diff
@@ -104,4 +105,112 @@ FunctionBuilderRegistry& FunctionBuilderRegistry::OnnxInstance() {
+  // Create a temporary unique node prefix for tensor names
+  std::string node_name = node.has_name() ?  node.name() : func.name() + 
+    std::to_string(fn_counter);
+  int version = (int)func.since_version();
+  std::unordered_map<std::string, std::string> input_names_map;
+  std::unordered_map<std::string, std::string> output_names_map;
+  std::unordered_map<std::string, AttributeProto> attr_map;
+
+  for (int idx = 0; idx < node.input_size(); ++idx) {
+    input_names_map[func.input().Get(idx)] = node.input().Get(idx);
+  }
+  for (int idx = 0; idx < node.output_size(); ++idx) {
+    output_names_map[func.output().Get(idx)] = node.output().Get(idx);
+  }
+
+  for (auto& attr : node.attribute()) {
+    attr_map[attr.name()] = attr;
+  }
+
+  for (auto& function_node : func.node()) {
+    NodeProto* new_node = g.add_node();
+    new_node->CopyFrom(function_node);
+    new_node->clear_input();
+    new_node->clear_output();
+    new_node->clear_attribute();
+    for (auto& input : function_node.input()) {
+      if (input_names_map.count(input)) {
+        new_node->add_input(input_names_map[input]);
+      } else {
+        new_node->add_input(RenameTensorNode(node_name, input));
+      }
+    }
+    for (auto& output : function_node.output()) {
+      if (output_names_map.count(output)) {
+        new_node->add_output(output_names_map[output]);
+      } else {
+        new_node->add_output(RenameTensorNode(node_name, output));
+      }
+    }
+    for (auto& attr : function_node.attribute()) {
+      if (attr.has_ref_attr_name()) {
+        if (attr_map.count(attr.ref_attr_name())) {
+          AttributeProto* new_attr = new_node->add_attribute();
+          new_attr->CopyFrom(attr_map[attr.ref_attr_name()]);
+        }
+      } else {
+        AttributeProto* new_attr = new_node->add_attribute();
+        new_attr->CopyFrom(attr);
+      }
+    }
+  }
+}
+
+Common::Status DecomposeGraph(
+    GraphProto& g,
+    const std::string& domain,
+    std::vector<std::string> function_list) {
+  GraphProto new_g = GraphProto(g);
+  const std::vector<OpSchema> op_schemas = OpSchemaRegistry::get_all_schemas();
+  std::unordered_set<std::string> registered_schemas;
+  for (const auto& op : op_schemas) {
+    registered_schemas.insert(op.Name());
+  }
+
+  std::multimap<std::string, std::unique_ptr<FunctionProto>> pfunction_map;
+  FunctionBuilderRegistry& function_registry =
+      FunctionBuilderRegistry::OnnxInstance();
+  Common::Status status =
+      function_registry.GetFunctions(domain, &pfunction_map);
+  std::unordered_map<std::string, size_t> fcounter_map;
+
+  new_g.clear_node();
+  std::unordered_set<std::string> function_set(
+      function_list.begin(), function_list.end());
+
+  for (int idx = 0; idx < g.node_size(); ++idx) {
+    auto& node = g.node().Get(idx);
+    if (!function_set.count(node.op_type()) &&
+        registered_schemas.count(node.op_type())) {
+      auto temp_node = new_g.add_node();
+      temp_node->CopyFrom(node);
```

Comment:
Can merge these two lines to be cleaner, or maybe add a parameter to add_node() to copy in one step?